File: src/compute/moment/anomaly_detector.py
from typing import Any
import numpy as np

class OmniResult:
    def __init__(self, value: Any, error: str = None):
        self.value = value
        self.error = error
        self.is_ok = error is None

class AnomalyDetector:
    def detect_spikes(self, timeseries: np.ndarray, threshold_std: float = 3.0) -> OmniResult:
        if timeseries is None or timeseries.size == 0:
            return OmniResult(None, "Empty time series")
            
        try:
            mean = np.mean(timeseries)
            std = np.std(timeseries)
            
            anomalies = np.abs(timeseries - mean) > (threshold_std * std)
            return OmniResult(anomalies)
        except Exception as e:
            return OmniResult(None, str(e))

File: src/business/timeseries/anomaly_detector.cs
using System;
using System.Collections.Generic;
using Omni.Core.Result;

namespace Omni.Business.TimeSeries
{
    // OMNI BUSINESS LAYER: Time Series Anomaly Detector
    // Applies Z-Score based anomaly detection over rolling windows.

    public class AnomalyEvent
    {
        public int Index { get; set; }
        public double Value { get; set; }
        public double ZScore { get; set; }
        public bool IsSpike { get; set; }
    }

    public class AnomalyDetector
    {
        public OmniResult<List<AnomalyEvent>, string> DetectAnomalies(List<double> series, double zThreshold = 3.0)
        {
            try
            {
                var anomalies = new List<AnomalyEvent>();
                if (series.Count < 2) return OmniResult<List<AnomalyEvent>, string>.Ok(anomalies);

                double sum = 0;
                foreach (var val in series) sum += val;
                double mean = sum / series.Count;

                double sumSq = 0;
                foreach (var val in series) sumSq += Math.Pow(val - mean, 2);
                double stdDev = Math.Sqrt(sumSq / (series.Count - 1));

                if (stdDev == 0) return OmniResult<List<AnomalyEvent>, string>.Ok(anomalies);

                for (int i = 0; i < series.Count; i++)
                {
                    double zScore = (series[i] - mean) / stdDev;
                    if (Math.Abs(zScore) > zThreshold)
                    {
                        anomalies.Add(new AnomalyEvent
                        {
                            Index = i,
                            Value = series[i],
                            ZScore = zScore,
                            IsSpike = zScore > 0
                        });
                    }
                }

                return OmniResult<List<AnomalyEvent>, string>.Ok(anomalies);
            }
            catch (Exception ex)
            {
                return OmniResult<List<AnomalyEvent>, string>.Err($"Anomaly detection failed: {ex.Message}");
            }
        }
    }
}
